### PR TITLE
implement multiple variable scope #48

### PR DESCRIPTION
**Pull request by lastchiliarch**
I'm try to implement multiple variable scope in orde to support input variable mapping in local scope.
   It contains some break change and has a lot codes, May I ask you to take some time to review ?

   
    Here is the main changes. 

    1. implement mulitply variable scope
    2. inherit parent scope when job is created
    3. introduce local scope
    4. modify job complete function  to support variable scope propagation
    5. remove embed processInfo from activieJob
    6. put input variable mapping into local scope

### CODE REVIEW

**Comment by nitram509** [comment]
Looks good at a first brief scan of mine.
I plan to review & merge by next week due to some private obligations.

**Comment by lastchiliarch** [comment]
> Looks good at a first brief scan of mine. I plan to review & merge by next week due to some private obligations.

Thank you very much.

File: pkg/bpmn_engine/engine.go
Comment (nitram509) [inline pkg/bpmn_engine/engine.go:53]: the field should still be named "variableContext". please rename it back.
the concept of scopes, only works with multiple variableContexts.

File: pkg/bpmn_engine/engine_jobs_activated.go
Comment (nitram509) [inline pkg/bpmn_engine/engine_jobs_activated.go:30]: This change does not relate to introducing variable context. Please remove it in this PR and create a separate PR, if required.
Also, this seems to be redundant, since there is already: ```GetProcessInstanceKey()``` available.

File: pkg/bpmn_engine/engine_jobs_activated.go
Comment (nitram509) [inline pkg/bpmn_engine/engine_jobs_activated.go:32]: This change does not relate to introducing variable context. Please remove it in this PR and create a separate PR, if required.

File: pkg/bpmn_engine/engine_jobs_activated.go
Comment (nitram509) [inline pkg/bpmn_engine/engine_jobs_activated.go:34]: This change does not relate to introducing variable context. Please remove it in this PR and create a separate PR, if required.

File: pkg/bpmn_engine/engine_jobs_activated.go
Comment (nitram509) [inline pkg/bpmn_engine/engine_jobs_activated.go:40]: Here I realized, we have a clash/issue in the current  ```main``` branch.
That's my fault. Accidently I let the interface ActivatedJob inherit from ProcessInstance,
but this makes no sense at all.
I need to refactor this. 
FYI: This GetVariable() function in the ActivatedJob interface will stay
No need to make a change from your side at this line.

File: pkg/bpmn_engine/engine_jobs_activated.go
Comment (nitram509) [inline pkg/bpmn_engine/engine_jobs_activated.go:43]: Setting variables just on the activatedJob interface is not good.
I consider we do support User Tasks and Service Tasks.
I would like to alter the process instance variables ONLY when a task is completed OR failed,
but never "in-between".
Please adjust, so that variables are set, when ```.Fail(vars)``` or ```.Complete(vars)``` is called.

File: pkg/bpmn_engine/engine_jobs_activated.go
Comment (nitram509) [inline pkg/bpmn_engine/engine_jobs_activated.go:46]: Reading the BPMN documentation from Camunda, I don't see a need, for introducing a local scope to lib-bpmn-engine users. This just gives more confusion. Please remove such.

File: pkg/bpmn_engine/engine_process_instance.go
Comment (nitram509) [inline pkg/bpmn_engine/engine_process_instance.go:16]: please keep 'variableContext' name

File: pkg/bpmn_engine/variable_scope/local_scope.go
Comment (nitram509) [inline pkg/bpmn_engine/variable_scope/local_scope.go:3]: the LocalScope implementation seems redundant to me.
Why not simply use ```localScope := variable_scope.NewScope(nil, nil)```?
So we could simplify the code?

File: pkg/bpmn_engine/variable_scope/scope.go
Comment (nitram509) [inline pkg/bpmn_engine/variable_scope/scope.go:3]: I like you put this into a separate package.
please don't use Abbreviations, but rather "VariableContext" instead.

File: pkg/bpmn_engine/variable_scope/scope_test.go
Comment (nitram509) [inline pkg/bpmn_engine/variable_scope/scope_test.go:18]: typo: Mulitiple ==> Multiple

File: pkg/bpmn_engine/variable_scope/scope_test.go
Comment (nitram509) [inline pkg/bpmn_engine/variable_scope/scope_test.go:67]: that's redundant and already covered by other Tests.
please remove.

File: pkg/bpmn_engine/engine_jobs_activated.go
Comment (nitram509) [inline pkg/bpmn_engine/engine_jobs_activated.go:22]: As I wrote in the other comment, an activated Job should not be able to alter its variables when in-flight ... but only when completing or failing. 
Also, a job only should contain one variable Context, not two.
Makes sense?

**Review by nitram509 — CHANGES_REQUESTED** [review]

**Comment by nitram509** [comment]
Hi @lastchiliarch 
I found the issue with altering the variable context when job are in flight, see #55 
I try to fix this by tomorrow, so you can merge it into your PR.

File: pkg/bpmn_engine/engine_jobs_activated.go
Comment (rhzs) [reply pkg/bpmn_engine/engine_jobs_activated.go:40]: Hi @nitram509 "That's my fault. Accidently I let the interface ActivatedJob inherit from ProcessInstance,
but this makes no sense at all.
I need to refactor this."

Yeah, I remember I put `ProcessInstance` inherit from `ActivatedJob`. I put that because I saw the previous code uses `ProcessInstance`.

File: pkg/bpmn_engine/engine.go
Comment (lastchiliarch) [reply pkg/bpmn_engine/engine.go:53]: I have renamed it back, thantks very much.

File: pkg/bpmn_engine/engine_jobs_activated.go
Comment (lastchiliarch) [reply pkg/bpmn_engine/engine_jobs_activated.go:40]: > Here I realized, we have a clash/issue in the current `main` branch. That's my fault. Accidently I let the interface ActivatedJob inherit from ProcessInstance, but this makes no sense at all. I need to refactor this. FYI: This GetVariable() function in the ActivatedJob interface will stay No need to make a change from your side at this line.


Yes, ActivatedJob  no need to  inherit  from ProcessInstance.

On the other hand, I thinks it's better to let each job has its own variable context than using the processInstance's context.

image

It's common to parallelize generata_update_msg and generata_alarm_msg , the task generata_update_msg will see the variable's middle status  in the execution of generata_alarm_msg task if some variable named name will be update servial times .
Of couse, it's a bad code style to relay on variable like this.

Another reason I like new variable context is to follow  camunda's doc which says:

"When the process instance enters a subprocess or an activity, a new scope is created. Activities in this scope can see all variables of this and of higher scopes (i.e. parent scopes). However, activities outside of this scope can not see the variables which are defined in this scope."

So I think it's better to let each job has its owne scope and propagated all varaible to its parente scope which maybe processInstance's scope when the job was completed.


That's  the reason I'm trying to introduce GetVariable function in ActivatedJob.
The new introduced GetVariable  function  will conflicted (or misleading) with the same function inherit   from ProcessInstance.


I will merge the code of main branch after you refactoring job.

File: pkg/bpmn_engine/variable_scope/scope.go
Comment (lastchiliarch) [reply pkg/bpmn_engine/variable_scope/scope.go:3]: Good idea.

File: pkg/bpmn_engine/engine_jobs_activated.go
Comment (lastchiliarch) [reply pkg/bpmn_engine/engine_jobs_activated.go:22]: Good idea, I'll remove one variable context.

**Comment by nitram509** [comment]
@lastchiliarch I took a bit longer than expected, but the interface is now cleaned up.

**Comment by lastchiliarch** [comment]
> @lastchiliarch I took a bit longer than expected, but the interface is now cleaned up.

Thanks for you help. I'm working on this.

File: pkg/bpmn_engine/engine_jobs_activated.go
Comment (lastchiliarch) [reply pkg/bpmn_engine/engine_jobs_activated.go]: > Setting variables just on the activatedJob interface is not good. I consider we do support User Tasks and Service Tasks. I would like to alter the process instance variables ONLY when a task is completed OR failed, but never "in-between". Please adjust, so that variables are set, when `.Fail(vars)` or `.Complete(vars)` is called.

What about leaving this part of job to issue 55 which was opened to trace it ?

File: pkg/bpmn_engine/engine_jobs_activated.go
Comment (lastchiliarch) [reply pkg/bpmn_engine/engine_jobs_activated.go]: Good point, it's not a good way to introduce SetVariableLocal. I have removed this function.

**Comment by lastchiliarch** [comment]
I have removed local scope and rebase the code to main,  could you please spend some time to review it again? @nitram509

**Comment by lastchiliarch** [comment]
> I have removed local scope and rebase the code to main, could you please spend some time to review it again? @nitram509

Seems it takes a long time on this pr,  could you have a look at this again? If anything need improve, I will work on this again. @nitram509

**Comment by nitram509** [comment]
Hi @lastchiliarch 
I'm sorry for the long pause.
I will take a look at your latest updates over the upcoming days.

**Comment by nitram509** [comment]
Hi @lastchiliarch 
I did review your PR and took your good ideas one step further.
I wanted to simplify the variable scope/context to just a unified implementation, not two 'VarScope' implementations.
With this unified implementation, the name changed to VariableHolder.
Also, I wanted the VariableHolder to be self-contained as much as possible, following encapsulation principles.
That means, the holder knows, what propagating means and does the copies on its own.

Here's what I came up with: https://github.com/nitram509/lib-bpmn-engine/pull/68/files
The implementation is shorter.

After all, I found it difficult to merge the meanwhile evolved main branch.
So, I will close this PR and cherry-pick to main manually.

Thank you very much for your patience and support.

**Pull request closed without merging**
